pyquil/operator_estimation.py

```diff
@@ -479,3 +541,93 @@ def measure_observables(qc: QuantumComputer, tomo_experiment: TomographyExperime
+    by keys of diagonal_sets.
+    Returns a different key if ``expt_setting`` isn't diagonal in the same tpb as any of the keys.
+    If it is diagonal in the same tpb as some key, then the corresponding value (a list of ExperimentSettings)
+    is updated to include ``expt_setting`` and the key is updated so it has the largest weight PauliTerm
+    specifying the tpb.
+
+    :param expt_setting: ExperimentSetting
+    :param diagonal_sets: dictionary keyed with ExperimentSetting, and with each value being a
+            list of ExperimentSettings
+    :return: the updated diagonal_sets dictionary updated with the input expt_setting placed
+            appropriately into it
+    """
+    # loop through dict items
+    for key, es_list in diagonal_sets.items():
+        # determine if key is diagonal in the same tpb as expt_setting
+        b_diagonal_tpb = _expt_settings_diagonal_in_tpb(key, expt_setting)
+        # update the dict value if so
+        if b_diagonal_tpb:
+            # determine the updated list of ExperimentSettings
+            updated_es_list = es_list + [expt_setting]
+            # obtain the diagonalizing bases for both the updated in and out sets
+            diagonalizing_in_term = _get_diagonalizing_basis([es.in_operator for es in updated_es_list])
+            diagonalizing_out_term = _get_diagonalizing_basis([es.out_operator for es in updated_es_list])
+            # update the diagonalizing basis (key of dict) if necessary
+            if len(diagonalizing_in_term) > len(key.in_operator) or len(diagonalizing_out_term) > len(key.out_operator):
+                del diagonal_sets[key]
+                new_key = ExperimentSetting(diagonalizing_in_term, diagonalizing_out_term)
+                diagonal_sets[new_key] = updated_es_list
+            else:
+                diagonal_sets[key] = updated_es_list
+            return diagonal_sets
+
+    # made it through entire dict without finding any ExperimentSetting with shared tpb,
+    # so need to make a new item
+    diagonal_sets[expt_setting] = [expt_setting]
+
+    return diagonal_sets
+
+
+def _commuting_sets_by_zbasis_tomo_expt(tomo_expt: TomographyExperiment):
+    """
+    Given an input TomographyExperiment, provide a dictionary indicating which ExperimentSettings
+    share a tensor product basis
+
+    :param tomo_expt: TomographyExperiment, from which to group ExperimentSettings that share a tpb
+        and can be run together
+    :return: dict with (key, value): ((diagonal_in_basis, diagonal_out_basis): list of tuples of PauliTerms
+                                                    that are diagonal in those bases)
+    """
+    diagonal_sets = {}
+    for expt_setting in tomo_expt:
+        assert len(expt_setting) == 1, 'already grouped?'
+        expt_setting = expt_setting[0]
+        diagonal_sets = _max_tpb_overlap(expt_setting, diagonal_sets)
+    return diagonal_sets
+
+
+def group_experiments_greedy(tomo_expt: TomographyExperiment):
```

Comment:
on the subject of grouping: try to organize the functions within this file:

 - similar / analogous functionality (ie the two "group" functions) should live near each other
 - try to order it so things only call functions defined above them culminating in the meaty user-facing functions at the bottom or at the bottom of a group of related functionality (use your judgement)

Reply:
I've re-organized as follows:

- At the very top sit the classes (ExperimentSetting, etc.) and any private helper functions they need
- The class definitions are followed by some functionality converting operator_estimation objects to/from JSONs
- In the middle is the user-facing function `measure_observables(..)` immediately preceded by its private helper functions
- Towards the end sit the user-facing functions for grouping, similarly preceded by private helper functions needed to define them